subprojects/core/src/integTest/groovy/org/gradle/api/tasks/FailingIncrementalTasksIntegrationTest.groovy

```diff
@@ -33,15 +34,83 @@ class FailingIncrementalTasksIntegrationTest extends AbstractIntegrationSpec {
             }
         """
 
+        expect:
+        succeeds "foo"
+
         when:
-        run "foo"
         file("out.txt") << "force rerun"
-        def failure1 = runAndFail "foo"
-        def failure2 = runAndFail "foo"
+        fails"foo"
+        then:
+        failureHasCause "Boo!"
 
+        when:
+        fails "foo", "--info"
         then:
-        failure1.assertHasCause("Boo!")
-        failure2.assertHasCause("Boo!")
+        failureHasCause "Boo!"
+        output.contains "Task has failed previously."
         //this exposes an issue we used to have with in-memory cache.
     }
+
+    @Unroll
+    def "incremental task after previous #failiureCount failure(s) #description"() {
```

Comment:
Say we have two input files, A and B. Let's say the following happens:
- task executes
- input A changed, task fail
- input A changed as in failed execution and input B changed, task succeeds

Then we would need to check that the incremental task receives files A and B as changed inputs and not only input B.